In [33]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

# Part 1: Load the dataset

In [34]:
# Load the given datasets
categorical_df = pd.read_csv('chronic_kidney_disease_categorical.csv')
numerical_df = pd.read_csv('chronic_kidney_disease_numerical.csv')
# Print the data
print(categorical_df.head())
print(numerical_df.head())

   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0    normal    normal  notpresent  notpresent   no   no   

  cad appet   pe  ane  Target  
0  no  good  yes   no     ckd  
1  no  good   no   no  notckd  
2  no  good  yes   no     ckd  
3  no  good   no  yes     ckd  
4  no  good   no   no  notckd  
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     76411

In [35]:
print(numerical_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 420 entries, 0 to 419
Data columns (total 12 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   unique_id  420 non-null    int64  
 1   age        412 non-null    float64
 2   bp         405 non-null    float64
 3   bgr        376 non-null    float64
 4   bu         401 non-null    float64
 5   sc         403 non-null    float64
 6   sod        331 non-null    float64
 7   pot        330 non-null    float64
 8   hemo       364 non-null    float64
 9   pcv        341 non-null    float64
 10  wbcc       305 non-null    float64
 11  rbcc       279 non-null    float64
dtypes: float64(11), int64(1)
memory usage: 39.5 KB
None


In [36]:
print(categorical_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 430 entries, 0 to 429
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   unique_id  430 non-null    int64  
 1   al         380 non-null    float64
 2   su         376 non-null    float64
 3   rbc        267 non-null    object 
 4   pc         360 non-null    object 
 5   pcc        426 non-null    object 
 6   ba         426 non-null    object 
 7   htn        428 non-null    object 
 8   dm         428 non-null    object 
 9   cad        428 non-null    object 
 10  appet      429 non-null    object 
 11  pe         429 non-null    object 
 12  ane        429 non-null    object 
 13  Target     430 non-null    object 
dtypes: float64(2), int64(1), object(11)
memory usage: 47.2+ KB
None


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The Chronic Kidney Disease (CKD) dataset from the UCI Machine Learning Repository consists of patient data, which helps in predicting the presence of chronic kidney disease based on various medical features. The dataset contains 25 attributes representing different health measurements or tests, including a target label indicating whether or not the patient has chronic kidney disease.
The dataset contains both categorical and numerical features.\
Examples of Categorical Features:\
    1. bp (Blood Pressure): Represented as a category like "normal" or "high".\
    2. al (Albumin): Denotes the presence of albumin in the urine, represented by levels 0-5.\
    3. ba (bacteria): indicates whether the bacteiral is present or not.\
Examples of numerical Features:\
    1. age: The patient’s age.\
    2. bp (Blood Pressure): indicates patient's bp in mm/Hg.\
    3. hemo (Hemoglobin): indicates patient's Hemoglobin in gms.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [37]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicate_numerical_rows = numerical_df.duplicated()
print("Total number of duplicate rows in the numerical dataset:", duplicate_numerical_rows.sum())
# Drop these duplicate rows
modified_numerical_df = numerical_df.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicate_categorical_rows = categorical_df.duplicated()
print("Total number of duplicate rows in the categorical dataset:", duplicate_categorical_rows.sum())

modified_categorical_df = categorical_df.drop_duplicates()

Total number of duplicate rows in the numerical dataset: 30
Total number of duplicate rows in the categorical dataset: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No, all the unique ids are not present in both datasets because, in the combined table after performing the outer join, the rows missing from the numerical dataset have missing numerical values (represented as NaN), but it contain categorical data from the categorical dataset. Therefore, we can infer that there are some rows in the categorical dataset with unique_ids that are missing in the numerical dataset resulting in NaN values for those rows in the numerical columns after the join.

In [38]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_dataset = pd.merge(modified_numerical_df, modified_categorical_df, how='outer', on='unique_id')
#Print the combined dataset
print(merged_dataset)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [39]:
# Calculate the percentage of rows that contain atleast one missing value
total_rows = merged_dataset.shape[0]
rows_with_missing_values = merged_dataset.isnull().any(axis=1).sum()
percentage_missing = (rows_with_missing_values / total_rows) * 100
# Print %
print("percentage of rows that contain atleast one missing value",percentage_missing)
# Drop these rows from the dataset
cleaned_merged_dataset = merged_dataset.dropna()
# Print the Dataset
print(cleaned_merged_dataset)

percentage of rows that contain atleast one missing value 61.75000000000001
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  

## Part 3.4: Sort the dataset according to the Labels

In [40]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_data = cleaned_merged_dataset.sort_values(by='Target')
sorted_data.reset_index(drop=True, inplace=True)
# Print the dataset
print(sorted_data)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [41]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
data_encoded = pd.get_dummies(sorted_data, columns=categorical_columns)
# Print the dataset
print(data_encoded)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: For binary variables, using a single column is equivalent and more efficient because the two columns are redundant (one is always the opposite of the other).

But for categorical variables with more than two values, one-hot encoding is preferred because it avoids creating an unintended ordinal relationship and allows the model to treat each category independently.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [42]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
cleaned_encoded_data = data_encoded.copy()
for column in numerical_columns:
    mean = cleaned_encoded_data[column].mean()
    std = cleaned_encoded_data[column].std()
    # Filter out the values.
    cleaned_encoded_data = cleaned_encoded_data[(cleaned_encoded_data[column] >= mean - 3 * std) & (cleaned_encoded_data[column] <= mean + 3 * std)]
# Print the dataset
print(cleaned_encoded_data)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
2       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
5       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
7       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
8       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
9       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [43]:
# Normalize the all Numerical Attributes in the dataset.
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
scaler = MinMaxScaler()
# Using .loc to avoid SettingWithCopyWarning
cleaned_encoded_data.loc[:, numerical_columns] = scaler.fit_transform(cleaned_encoded_data[numerical_columns])
# Print the dataset
print(cleaned_encoded_data)

     unique_id       age    bp       bgr        bu        sc       sod  \
2       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
5       397388  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
7       438182  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
8       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
9       475200  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..         ...       ...   ...       ...       ...       ...       ...   
148     486397  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
149     203466  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
150     482105  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
151     228849  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
152     996076  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_good  \
2    0.793103  0.000000  0.032258  ...  

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: From the dataset, the column unique_id appears to be an identifier column that is likely not useful for modeling or predictions because it does not provide any meaningful information about the relationship between features and the target variable. Therefore, This column should be removed.

In [44]:
#Remove that column
cleaned_encoded_data = cleaned_encoded_data.drop(columns=['unique_id'])
# Print the dataset
print(cleaned_encoded_data)

          age    bp       bgr        bu        sc       sod       pot  \
2    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
5    0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
7    0.905405  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
8    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
9    0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.189189  0.25  0.227468  0.222222  0.074074  0.500000  0.310345   
149  0.216216  0.00  0.115880  0.052288  0.098765  0.600000  0.482759   
150  0.513514  0.00  0.107296  0.235294  0.012346  0.600000  0.448276   
151  0.702703  0.50  0.128755  0.261438  0.098765  0.666667  0.206897   
152  0.689189  0.00  0.150215  0.254902  0.098765  1.000000  0.620690   

         hemo       pcv      wbcc  ...  cad_no  cad_yes  appet_good  \
2    0.000000  0.032258  0.395161  ...   False     T

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [45]:
# Export the dataframe to a new csv file
cleaned_encoded_data.to_csv('/workspaces/assignment-3-data-cleaning-and-preprocessing-<redacted>/cleaned_encoded_data.csv', index=False)
# Export the dataframe to a new json file
cleaned_encoded_data.to_json('/workspaces/assignment-3-data-cleaning-and-preprocessing-<redacted>/cleaned_encoded_data.json', orient='records', lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [46]:
fifteen_data = cleaned_encoded_data.head(15)
print(fifteen_data)

         age    bp       bgr        bu        sc       sod       pot  \
2   0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
5   0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
7   0.905405  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
8   0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
9   0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
10  0.729730  0.00  0.935622  0.169935  0.160494  0.333333  0.034483   
12  0.675676  0.75  0.253219  0.633987  0.777778  0.366667  0.655172   
13  0.851351  0.25  0.618026  0.562092  0.728395  0.000000  0.344828   
15  0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
18  0.756757  0.25  0.223176  0.209150  0.160494  0.533333  0.620690   
19  0.662162  0.50  0.618026  0.411765  0.432099  0.566667  0.689655   
20  0.783784  0.00  0.725322  0.313725  0.481481  0.566667  0.862069   
22  0.878378  0.00  0.206009  0.751634  0.604938  0.533333  0.68

### Paste the markdown table here
|     |     age |   bp |      bgr |      bu |      sc |      sod |       pot |       hemo |       pcv |      wbcc | ... | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|-----|---------|------|----------|---------|---------|----------|-----------|------------|-----------|-----------|-----|--------|---------|------------|-------------|-------|--------|--------|---------|-------------|----------------|
|  2  | 0.743243| 0.50 | 0.442060 | 0.901961| 0.432099| 0.500000 | 0.793103  |  0.000000  | 0.032258  | 0.395161  | ... | False  |   True  |   False    |     True    | False |  True  | False  |    True |     True    |      False      |
|  5  | 0.770270| 1.00 | 0.901288 | 0.163399| 0.345679| 0.766667 | 0.206897  |  0.524752  | 0.548387  | 0.443548  | ... | False  |   True  |    True    |    False    |  True |  False |  True  |   False |     True    |      False      |
|  7  | 0.905405| 0.50 | 0.785408 | 0.862745| 0.518519| 0.600000 | 1.000000  |  0.277228  | 0.322581  | 0.233871  | ... | False  |   True  |    True    |    False    |  True |  False |  True  |   False |     True    |      False      |
|  8  | 0.202703| 0.75 | 0.158798 | 0.196078| 0.160494| 0.166667 | 0.206897  |  0.059406  | 0.000000  | 0.653226  | ... |  True  |  False  |    True    |    False    |  True |  False | False  |    True |     True    |      False      |
|  9  | 0.783784| 1.00 | 0.399142 | 0.287582| 0.839506| 0.666667 | 0.586207  |  0.019802  | 0.096774  | 0.258065  | ... |  True  |  False  |    True    |    False    | False |  True  |  True  |   False |     True    |      False      |
| 10  | 0.729730| 0.00 | 0.935622 | 0.169935| 0.160494| 0.333333 | 0.034483  |  0.019802  | 0.064516  | 0.879032  | ... |  True  |  False  |   False    |     True    |  True |  False | False  |    True |     True    |      False      |
| 12  | 0.675676| 0.75 | 0.253219 | 0.633987| 0.777778| 0.366667 | 0.655172  |  0.138614  | 0.193548  | 0.169355  | ... |  True  |  False  |    True    |    False    |  True |  False |  True  |   False |     True    |      False      |
| 13  | 0.851351| 0.25 | 0.618026 | 0.562092| 0.728395| 0.000000 | 0.344828  |  0.168317  | 0.161290  | 0.580645  | ... | False  |   True  |    True    |    False    |  True |  False |  True  |   False |     True    |      False      |
| 15  | 0.540541| 0.00 | 0.399142 | 0.535948| 0.358025| 0.700000 | 0.379310  |  0.207921  | 0.161290  | 0.830645  | ... |  True  |  False  |    True    |    False    |  True |  False |  True  |   False |     True    |      False      |
| 18  | 0.756757| 0.25 | 0.223176 | 0.209150| 0.160494| 0.533333 | 0.620690  |  0.485149  | 0.516129  | 0.290323  | ... |  True  |  False  |    True    |    False    |  True |  False |  True  |   False |     True    |      False      |
| 19  | 0.662162| 0.50 | 0.618026 | 0.411765| 0.432099| 0.566667 | 0.689655  |  0.316832  | 0.354839  | 0.250000  | ... |  True  |  False  |    True    |    False    |  True |  False |  True  |   False |     True    |      False      |
| 20  | 0.783784| 0.00 | 0.725322 | 0.313725| 0.481481| 0.566667 | 0.862069  |  0.178218  | 0.193548  | 0.258065  | ... |  True  |  False  |   False    |     True    | False |  True  |  True  |   False |     True    |      False      |
| 22  | 0.878378| 0.00 | 0.206009 | 0.751634| 0.604938| 0.533333 | 0.689655  |  0.366337  | 0.387097  | 0.879032  | ... |  True  |  False  |   False    |     True    | False |  True  |  True  |   False |     True    |      False      |
| 23  | 0.878378| 0.25 | 0.639485 | 0.470588| 0.395062| 0.433333 | 0.517241  |  0.267327  | 0.322581  | 0.104839  | ... | False  |   True  |    True    |    False    |  True |  False |  True  |   False |     True    |      False      |
| 29  | 0.716216| 0.50 | 1.000000 | 0.163399| 0.111111| 0.066667 | 0.206897  |  0.267327  | 0.387097  | 0.532258  | ... |  True  |  False  |   False    |     True    |  True |  False |  True  |   False |     True    |      False      |



** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer: When converting datasets between formats (e.g., CSV to JSON), we can rely on specialized tools like pandas in Python, which are designed to handle these tasks accurately. For instance, we can use pandas.to_csv() and pandas.to_json() for conversions instead of relying on LLMs for large datasets. Moreover, after conversion, we can validate the output using traditional coding tools. For example, check the structure, missing values, or even row/column counts to ensure the conversion didn't have any errors.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [47]:
# Code to rename all the columns in the dataset
column_mapping = {
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the columns in the DataFrame
cleaned_encoded_data.rename(columns=column_mapping, inplace=True)
print(cleaned_encoded_data.head())

        age  blood pressure  blood glucose random  blood urea  \
2  0.743243            0.50              0.442060    0.901961   
5  0.770270            1.00              0.901288    0.163399   
7  0.905405            0.50              0.785408    0.862745   
8  0.202703            0.75              0.158798    0.196078   
9  0.783784            1.00              0.399142    0.287582   

   serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
2          0.432099  0.500000   0.793103    0.000000            0.032258   
5          0.345679  0.766667   0.206897    0.524752            0.548387   
7          0.518519  0.600000   1.000000    0.277228            0.322581   
8          0.160494  0.166667   0.206897    0.059406            0.000000   
9          0.839506  0.666667   0.586207    0.019802            0.096774   

       wbcc  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  \
2  0.395161  ...   False     True       False        True  False    True   


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [48]:
# Converted SQL to Pandas code
sorted_data_grouped = sorted_data.groupby('Target').size().reset_index(name='count')
print(sorted_data_grouped)

   Target  count
0     ckd     42
1  notckd    111


*Describe what the above code does here*

The code is performing a group-wise count operation, similar to the SQL query that groups by Target and counts how many records exist for each unique value in that column. In this case there are total 42 values for ckd and 111 values for notckd.